pkg/datarepair/datarepair_test.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package datarepair_test
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite"
+	"storj.io/storj/uplink"
+)
+
+func TestDataRepair(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount:   1,
+		StorageNodeCount: 10,
+		UplinkCount:      1,
+		Reconfigure: testplanet.Reconfigure{
+			Satellite: func(log *zap.Logger, index int, config *satellite.Config) {
+				config.Checker.Interval = 100 * time.Millisecond
+				config.Repairer.Interval = 100 * time.Millisecond
+			},
+		},
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+		// first, upload some remote data
+		ul := planet.Uplinks[0]
+		satellite := planet.Satellites[0]
+
+		testData := make([]byte, 1*memory.MiB)
+		_, err := rand.Read(testData)
+		assert.NoError(t, err)
+
+		err = ul.UploadWithConfig(ctx, satellite, &uplink.RSConfig{
+			MinThreshold:     2,
+			RepairThreshold:  3,
+			SuccessThreshold: 4,
+			MaxThreshold:     5,
+		}, "testbucket", "test/path", testData)
+		require.NoError(t, err)
+
+		// get a remote segment from pointerdb
+		pdb := satellite.Metainfo.Service
+		listResponse, _, err := pdb.List("", "", "", true, 0, 0)
+		require.NoError(t, err)
+
+		var path string
+		var pointer *pb.Pointer
+		for _, v := range listResponse {
+			path = v.GetPath()
+			pointer, err = pdb.Get(path)
+			assert.NoError(t, err)
+			if pointer.GetType() == pb.Pointer_REMOTE {
+				break
+			}
+		}
+
+		// calculate how many storagenodes to kill
+		numStorageNodes := len(planet.StorageNodes)
+		redundancy := pointer.GetRemote().GetRedundancy()
+		remotePieces := pointer.GetRemote().GetRemotePieces()
+		minReq := redundancy.GetMinReq()
+		numPieces := len(remotePieces)
+		toKill := numPieces - int(minReq)
+		// we should have enough storage nodes to repair on
+		assert.True(t, (numStorageNodes-toKill) >= numPieces)
+
+		// kill nodes and track lost pieces
+		var lostPieces []int32
+		nodesToKill := make(map[storj.NodeID]bool)
+		nodesToKeepAlive := make(map[storj.NodeID]bool)
+		for i, piece := range remotePieces {
+			if i >= toKill {
+				nodesToKeepAlive[piece.NodeId] = true
+				continue
+			}
+			nodesToKill[piece.NodeId] = true
+			lostPieces = append(lostPieces, piece.GetPieceNum())
+		}
+		for _, node := range planet.StorageNodes {
+			if nodesToKill[node.ID()] {
+				err = planet.StopPeer(node)
+				assert.NoError(t, err)
+			}
+		}
+
+		time.Sleep(2 * time.Second)
```

Comment:
is there anything we can do to avoid this sleep? does testplanet have a way to wait for the condition we really care about here?

Reply:
Most probably satellite `satellite.Repair.Checker.Loop` with methods `Pause()` and `TriggerWait()` can be used. With those methods we can control when specific operation will be triggered. Also `Loop` needs to be added to `repairer.Service`.

Example: https://github.com/storj/storj/blob/b05cf05649d053360f22bd1e8dac7da91036bac5/storagenode/monitor/monitor_test.go#L31

Reply:
I added `Loop` to `repairer.Service`, but `TriggerWait` does not wait for the repairer to finish. I will try to see if I can implement `Loop` differently to get around this issue

Reply:
Yeah, the problem is that repairing is starting work in goroutine and triggerWait is waiting only to finish `process` method.
https://github.com/storj/storj/blob/5cd7536cb689b243bc55dffb5e0cca72de9e75c4/pkg/datarepair/repairer/repairer.go#L101

Reply:
Fixed by exposing limiter and using `Limiter.Wait()`

Reply:
Fixed by exposing limiter and using its `.Wait()`